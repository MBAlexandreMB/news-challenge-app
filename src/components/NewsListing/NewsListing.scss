@import '../../shared/styles/sizes.scss';

.news-listing-container {
    width: fit-content;
}

@media (max-width: #{$breakpoint-tablet}) {
    .news-listing-container.selected {
        width: 100%;
        display: flex;
        flex-flow: row nowrap;
        overflow: scroll;
        overflow-y: hidden;
        scroll-behavior: smooth;
        gap: 10px;
    }
}