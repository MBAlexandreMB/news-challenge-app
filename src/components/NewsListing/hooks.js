import { useState, useEffect } from 'react';
import { getNews } from './api';

const useGetNews = () => {
    const [news, setNews] = useState([]);
    const [newsError, setNewsError] = useState(null);

    useEffect(() => {
        getNews(setNews, setNewsError);
    }, [setNews, setNewsError]);

    return [news, newsError];
};

export { useGetNews };