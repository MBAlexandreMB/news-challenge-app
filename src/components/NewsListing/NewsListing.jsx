import React from 'react';

import { useGetNews } from './hooks';
import NewsCard from './NewsCard/NewsCard';

const NewsListing = () => {
    const [newsList, newsListError] = useGetNews();

    return (
        <div>
            {
                newsList.length > 0 &&
                newsList.map(newsInfo =>  <NewsCard key={newsInfo.url} news={newsInfo} />)
            }
        </div>
    );
};
 
export default NewsListing;