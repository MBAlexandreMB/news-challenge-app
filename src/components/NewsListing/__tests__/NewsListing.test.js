import { shallow } from 'enzyme';
import { act } from 'react-dom/test-utils';
import NewsListing from '../NewsListing';

const newsList = [
    { url: 'fakeurl', title: 'faketitle' },
    { url: 'fakeurl2', title: 'faketitle2' },
];

const mockScheduleASearch = { start: jest.fn() };
const mockLoadMoreNews = jest.fn();

jest.mock('../../../hooks', () => ({
    useGetNews: () => ([newsList, mockLoadMoreNews, mockScheduleASearch]),
}));

jest.mock('react-intersection-observer', () => ({
    ...jest.requireActual('react-intersection-observer')
}));

const defaultProps = {
    isAnyNewsSelected: false,
    isLoading: false,
    onSelectNews: jest.fn(),
    setIsLoading: jest.fn(),
    onNewsLoaded: jest.fn(),
};

describe('<NewsListing />', () => {
    it('should render correctly', () => {
        const wrapper = shallow(<NewsListing {...defaultProps} />);

        expect(wrapper).toHaveLength(1);
        expect(wrapper).toMatchSnapshot();
    });

    it('should render correctly when loading', () => {
        const wrapper = shallow(<NewsListing {...defaultProps} isLoading={true} />);
        const loader = wrapper.find('Loader');
        
        expect(loader.exists()).toBe(true);
        expect(wrapper).toHaveLength(1);
        expect(wrapper).toMatchSnapshot();
    });


    it('should set news list class as selected when any news is selected', () => {
        const wrapper = shallow(<NewsListing {...defaultProps} isAnyNewsSelected={true} />);

        const isNewsListClassSelected = wrapper.find('[data-test="news-list"]').hasClass('selected');
        expect(isNewsListClassSelected).toBe(true);
    });

    it('should NOT set news list class as selected when no news is selected', () => {
        const wrapper = shallow(<NewsListing {...defaultProps} isAnyNewsSelected={false} />);

        const isNewsListClassSelected = wrapper.find('[data-test="news-list"]').hasClass('selected');
        expect(isNewsListClassSelected).toBe(false);
    });

    it('should schedule a search when search input changes', () => {
        const wrapper = shallow(<NewsListing {...defaultProps} />);

        jest.clearAllMocks();
        jest.useFakeTimers();
        act(() => {
            wrapper.find('[data-test="search-input"]').simulate('change', { target: { value: 'search value' } });
            jest.runAllTimers();
        })

        expect(mockScheduleASearch.start).toHaveBeenCalled();
    });
});