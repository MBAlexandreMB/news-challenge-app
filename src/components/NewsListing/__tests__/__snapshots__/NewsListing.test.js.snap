// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewsListing /> should render correctly 1`] = `
<section
  className="news-listing-container"
>
  <input
    className="search-input"
    data-test="search-input"
    onChange={[Function]}
    placeholder="Search"
    type="text"
    value=""
  />
  <ul
    className="news-listing-ul "
    data-test="news-list"
  >
    <li
      key="fakeurl"
    >
      <NewsCard
        key="fakeurl"
        news={
          Object {
            "title": "faketitle",
            "url": "fakeurl",
          }
        }
        onSelectNews={[MockFunction]}
      />
    </li>
    <li
      key="fakeurl2"
    >
      <NewsCard
        key="fakeurl2"
        news={
          Object {
            "title": "faketitle2",
            "url": "fakeurl2",
          }
        }
        onSelectNews={[MockFunction]}
      />
    </li>
  </ul>
</section>
`;

exports[`<NewsListing /> should render correctly when loading 1`] = `
<section
  className="news-listing-container"
>
  <input
    className="search-input"
    data-test="search-input"
    onChange={[Function]}
    placeholder="Search"
    type="text"
    value=""
  />
  <ul
    className="news-listing-ul "
    data-test="news-list"
  >
    <li
      key="fakeurl"
    >
      <NewsCard
        key="fakeurl"
        news={
          Object {
            "title": "faketitle",
            "url": "fakeurl",
          }
        }
        onSelectNews={[MockFunction]}
      />
    </li>
    <li
      key="fakeurl2"
    >
      <NewsCard
        key="fakeurl2"
        news={
          Object {
            "title": "faketitle2",
            "url": "fakeurl2",
          }
        }
        onSelectNews={[MockFunction]}
      />
    </li>
    <li
      className="cards-loading news-card-container"
    >
      <Loader
        borderSize="4px"
        size={24}
      />
    </li>
  </ul>
</section>
`;
